### ISSUE TITLE
[Exercises] Deleting the Only Iteration on an Exercise Leads to Strange State

### ISSUE THREAD

**Issue by BethanyG** [exercism/v3-beta#168]
Working with an exercise that has been both completed and published, but has only one iteration.

1.  Visit the exercise page, and click the [iteration] section/tab.
2.  Click the [...] in the Analyzer/Tests area and select [Delete Iteration]

The following displays:

![image](https://user-images.githubusercontent.com/5923094/128568161-2098359d-14a5-4fbc-b7b8-b665b331d99d.png)

3. Select [Delete Iteration]

The following displays:
![image](https://user-images.githubusercontent.com/5923094/128568542-2834918c-39d2-4694-a0f3-31fb437fbf19.png)

🔴   -expected a warning about the solution no longer being published (_there is no iteration **to** publish_).
🔴   -expected a warning about the exercise no longer being completed/answered (_there is no iteration to base those states on_).

4. Return to the exercise overview.

The following displays:
![image](https://user-images.githubusercontent.com/5923094/128568746-99b25240-98c5-431c-b60a-43f673788b17.png)

🔴  -expected to not have a "completed" message.
🔴  -expected to not have a "Your published solution" message/section

5.  Click on the "published solution" box
The following displays:
![image](https://user-images.githubusercontent.com/5923094/128568938-0cf71400-2f41-4457-b199-87b549231212.png)

🔴  -expected a message about there being no published solution, or better yet -- this page would not exist.

From here, the state feels more confused.  If I now return to the exercise and "open in editor", the deleted iteration shows back up.  If I then "submit", the result is now two iterations, not one.  While this makes sense from an app perspective (_after all, there are really two iterations ... we just deleted one of them_), it feels a bit weird from a user perspective.

![image](https://user-images.githubusercontent.com/5923094/128569419-5d741c60-86c8-46f0-a1c7-de7c9f6dfebe.png)

**Comment by iHiD** [comment]
Thanks 🙂 

---

> expected a warning about the solution no longer being published (there is no iteration to publish).

Agree with this. If there's no non-deleted iterations, then the solution should be unpublished (and there should be a warning as such)

>  -expected a warning about the exercise no longer being completed/answered (there is no iteration to base those states on).

I think the exercise stays completed, even though the iteration has gone. Because it will have unlocked the next exercises (and had other side effects) when completing, that is a one-way action which can't be reversed. It's weird to have a completed solution with no iterations (why would someone choose to get themselves into that state?), but I don't think it's wrong per se.

> From here, the state feels more confused. If I now return to the exercise and "open in editor", the deleted iteration shows back up. If I then "submit", the result is now two iterations, not one. While this makes sense from an app perspective (after all, there are really two iterations ... we just deleted one of them), it feels a bit weird from a user perspective.

Again, I think, while weird, this does need to be the case, as there might already be mentor comments etc attached to that first iteration. I'm not sure if there's a better way we can communicate this or not, but I think functionally I'm ok with this.

As a side note, I **think** the editor is actually showing the last code you wrote in it, as opposed to the last iteration. I think if you uploaded a new, bad iteration from the CLI and then deleted it, you'd find it doesn't show in the editor. At least that's the intended flow.

---

As an overarching thought, I'm presuming people are generally only going to delete "wrong" iterations (e.g. wrong file uploaded, accidently leaked password, etc). So I think I'm ok with things being a little "off" when that happens.

**Comment by BethanyG** [comment]
>  It's weird to have a completed solution with no iterations (why would someone choose to get themselves into that state?), but I don't think it's wrong per se.

Thinking someone could come back after a pause and get embarrassed or otherwise think what they completed no longer represents where they're at, and so want to "erase" it....and not be in the head space of uploading something else first.  But totally agree it wouldn't be a common thing.  And it wouldn't be a "wrong" state -- as long as there isn't a blank "published" solution.

> As a side note, I think the editor is actually showing the last code you wrote in it, as opposed to the last iteration. I think if you uploaded a new, bad iteration from the CLI and then deleted it, you'd find it doesn't show in the editor. At least that's the intended flow.

Ah!  yup.  😅

**Comment by iHiD** [comment]
> Thinking someone could come back after a pause and get embarrassed or otherwise think what they completed no longer represents where they're at, and so want to "erase" it....and not be in the head space of uploading something else first. 

Yeah - this makes sense. I guess because they could always unpublish it to make it private I'm less worried that people will feel the need to do this. But I know people do (I know people often restart tracks entirely for this reason for example).

I'll rename this to the first issue so I remember what it's about without having to reread :)
